lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -113,12 +113,13 @@ def serializable_hash_for_single_resource
         end
 
         def resource_identifier_type_for(serializer)
-          return serializer._type if serializer._type
-          if ActiveModelSerializers.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
-          else
-            serializer.object.class.model_name.plural
+          value_or_block = serializer._type
+          if value_or_block
+            return serializer.instance_eval(&value_or_block) if value_or_block.respond_to?(:call)
+            return value_or_block
           end
+          inflection = ActiveModelSerializers.config.jsonapi_resource_type
+          serializer.object.class.model_name.public_send(inflection)
```

Comment:
The structure of this code change makes me a bit uncomfortable. Since the change is that `type` is a value or a callable, I'd think that either the value is also made callable in the setter, or that it is consumed with a respond_to?(:call) and no more-- I'm surprised to see so many changes both here and in the tests.  I'm going to take a closer look when I have some time.  Would you mind sharing your design thoughts?